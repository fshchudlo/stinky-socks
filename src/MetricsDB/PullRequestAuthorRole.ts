export enum PullRequestAuthorRole {
    /** The author is the owner of the repository. */
    OWNER = "OWNER",

    /** The author is a member of the organization that owns the repository. */
    MEMBER = "MEMBER",

    /** The author has write access to the repository. */
    COLLABORATOR = "COLLABORATOR",

    /** The author is making their first contribution to any repository in the organization. */
    FIRST_TIMER = "FIRST_TIMER",

    /** The author is contributing to this particular repository for the first time. */
    FIRST_TIME_CONTRIBUTOR = "FIRST_TIME_CONTRIBUTOR",

    /** The author has previously committed to the repository but is not necessarily a member or collaborator. */
    CONTRIBUTOR = "CONTRIBUTOR",

    /** The author is a placeholder for a previously deleted user. */
    MANNEQUIN = "MANNEQUIN",

    /** The author has no affiliation with the repository. */
    NONE = "NONE"
}