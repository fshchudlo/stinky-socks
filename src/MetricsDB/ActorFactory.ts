import { Actor } from "./entities/Actor";
import { MetricsDB } from "./MetricsDB";
import { Repository } from "typeorm";
import { createCache } from "cache-manager";

const colors = [
    "Red", "Orange", "Yellowish", "Greenish", "Bluish", "Purplish", "Pinkish", "Brownish", "Blackish", "Whitish",
    "Grayish", "Teal", "Turquoise", "Magenta", "Lavender", "Coral", "Golden", "Silvery", "Beige", "Ivory", "Indigo",
    "Violet", "Minty", "Crimson", "Olive", "Maroon", "Navy", "Amber", "Cyan", "Azure", "Lime", "Cobalt", "Copper"
];

const adjectives: string[] = [
    "Bubbly", "Jazzy", "Snazzy", "Zesty", "Wobbly", "Dandy", "Doodle", "Loopy", "Perky",
    "Wiggly", "Sassy", "Peppy", "Swirly", "Fizzy", "Chipper", "Zippy", "Poppy", "Plucky",
    "Sprightly", "Mellow", "Jumpy", "Giddy", "Cheeky", "Spiffy", "Frothy", "Breezy", "Nifty",
    "Brave", "Clever", "Witty", "Kind", "Fierce", "Happy", "Jolly", "Sneaky", "Bouncy", "Dizzy",
    "Goofy", "Lucky", "Fluffy", "Cheerful", "Grumpy", "Wacky", "Puzzled", "Eccentric", "Spunky",
    "Quirky", "Funky", "Chirpy", "Snappy", "Whizzy", "Zany"
];

const animals = [
    "Lion", "Tiger", "Bear", "Wolf", "Eagle", "Shark", "Panda", "Giraffe", "Koala", "Penguin",
    "Sloth", "Otter", "Kangaroo", "Raccoon", "Squirrel", "Turtle", "Llama", "Monkey", "Hippo", "Elephant",
    "Rabbit", "Hedgehog", "Parrot", "Owl", "Moose", "Duck", "Goose", "Ferret", "Octopus", "Platypus"
];
const actorsRepository: Repository<Actor> = MetricsDB.getRepository(Actor);
const actorsCache: ReturnType<typeof createCache> = createCache();

export class ActorFactory {
    public static async preloadCacheByTeam(teamName: string): Promise<void> {
        const users = await actorsRepository.find({ where: { teamName: teamName } });
        for (const user of users) {
            const cacheKey = `${teamName}-${user.login}`;
            await actorsCache.set(cacheKey, user);
        }
    }

    public static async fetch({ teamName, login, isBotUser }: {
        teamName: string,
        login: string,
        isBotUser: boolean
    }): Promise<Actor> {
        const cacheKey = `${teamName}-${login}`;

        return await actorsCache.wrap(cacheKey, async () => {
            let user = await actorsRepository.findOne({ where: { teamName, login } });

            if (user) {
                return user;
            }

            const nickname = await ActorFactory.generateUniqueNickname(teamName);
            user = actorsRepository.create({
                teamName,
                login,
                isBotUser,
                isFormerParticipant: false,
                nickname
            });

            await actorsRepository.save(user);
            return user;
        });
    }

    private static async generateUniqueNickname(teamName: string): Promise<string> {
        let nickname: string;
        let isUnique = false;

        do {
            nickname = ActorFactory.generateNickname();
            const existingUser = await actorsRepository.findOne({ where: { teamName, nickname } });
            if (!existingUser) {
                isUnique = true;
            } else {
                console.log(`ðŸ‘» Nickname "${nickname}" is already taken. Checking for another one...`);
            }
        } while (!isUnique);

        return nickname;
    }

    private static generateNickname(): string {
        const color = colors[Math.floor(Math.random() * colors.length)];
        const adjective = adjectives[Math.floor(Math.random() * colors.length)];
        const animal = animals[Math.floor(Math.random() * animals.length)];
        return `${color} ${adjective} ${animal}`;
    }

}