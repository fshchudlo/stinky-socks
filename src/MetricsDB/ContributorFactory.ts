import { Contributor } from "./Contributor";
import { MetricsDB } from "./MetricsDB";
import { Repository } from "typeorm";
import { createCache } from "cache-manager";
import { AppConfig } from "../app.config";

const adjectives = [
    "Brave", "Clever", "Witty", "Kind", "Fierce", "Happy", "Jolly", "Sneaky", "Bouncy", "Dizzy",
    "Goofy", "Zippy", "Lucky", "Fluffy", "Cheerful", "Grumpy", "Wacky", "Puzzled", "Eccentric", "Spunky",
    "Quirky", "Funky", "Chirpy", "Nifty", "Snappy", "Peppy", "Perky", "Whizzy", "Zany", "Breezy"
];

const animals = [
    "Lion", "Tiger", "Bear", "Wolf", "Eagle", "Shark", "Panda", "Giraffe", "Koala", "Penguin",
    "Sloth", "Otter", "Kangaroo", "Raccoon", "Squirrel", "Turtle", "Llama", "Monkey", "Hippo", "Elephant",
    "Rabbit", "Hedgehog", "Parrot", "Owl", "Moose", "Duck", "Goose", "Ferret", "Octopus", "Platypus"
];
const contributorRepo: Repository<Contributor> = MetricsDB.getRepository(Contributor);
const contributorsCache = createCache();

export class ContributorFactory {
    public static async preloadCacheByTeam(teamName: string): Promise<void> {
        const contributors = await contributorRepo.find({ where: { teamName: teamName } });
        for (const contributor of contributors) {
            const cacheKey = `${teamName}-${contributor.login}`;
            await contributorsCache.set(cacheKey, contributor);
        }
    }

    public static async fetchContributor({ teamName, login, isBotUser, isFormerEmployee }: {
        teamName: string,
        login: string,
        isBotUser: boolean,
        isFormerEmployee: boolean
    }): Promise<Contributor> {
        const cacheKey = `${teamName}-${login}`;
        login = AppConfig.userNameNormalizerFn(login);

        return await contributorsCache.wrap(cacheKey, async () => {
            let contributor = await contributorRepo.findOne({ where: { teamName: teamName, login } });

            if (contributor) {
                return contributor;
            }

            const nickname = await ContributorFactory.generateUniqueNickname(teamName);
            contributor = contributorRepo.create({
                teamName,
                login,
                isBotUser,
                isFormerEmployee,
                nickname
            });

            await contributorRepo.save(contributor);
            return contributor;
        });
    }

    private static async generateUniqueNickname(team: string): Promise<string> {
        let nickname: string;
        let isUnique = false;

        do {
            nickname = ContributorFactory.generateNickname();
            const existingContributor = await contributorRepo.findOne({ where: { teamName: team, nickname } });
            if (!existingContributor) {
                isUnique = true;
            }
        } while (!isUnique);

        return nickname;
    }

    private static generateNickname(): string {
        const adjective = adjectives[Math.floor(Math.random() * adjectives.length)];
        const animal = animals[Math.floor(Math.random() * animals.length)];
        return `${adjective} ${animal}`;
    }

}