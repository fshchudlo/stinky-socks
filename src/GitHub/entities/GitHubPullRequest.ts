import { PullRequest } from "../../MetricsDB/entities/PullRequest";
import { GitHubPullRequestParticipant } from "./GitHubPullRequestParticipant";
import { UserFactory } from "../../MetricsDB/UserFactory";
import { ActivityTraits } from "./helpers/ActivityTraits";
import { ImportParams } from "./ImportParams";
import getCommentsTimestamps from "./helpers/getCommentsTimestamps";
import getActivitiesOf from "./helpers/getActivitiesOf";
import calculatePrSharedForReviewDate from "./helpers/calculatePrSharedForReviewDate";

export class GitHubPullRequest extends PullRequest {
    public async init(model: ImportParams) {
        await this.initializeBaseProperties(model);

        await this.initializeDates(model)
            .calculateCommitStats(model)
            .initializeParticipants(model);

        return this.validateDataIntegrity().calculateTimings();
    }

    private async initializeBaseProperties(model: ImportParams) {
        this.teamName = model.teamName;
        this.projectName = model.pullRequest.base.repo.owner.login;
        this.repositoryName = model.pullRequest.base.repo.name;
        this.pullRequestNumber = model.pullRequest.number;
        this.viewURL = model.pullRequest.html_url;
        this.targetBranch = model.pullRequest.base.ref;
        this.requestedReviewersCount = model.pullRequest.requested_reviewers.length;
        this.authorRole = model.pullRequest.author_association;
        this.createdDate = new Date(model.pullRequest.created_at);
        this.updatedDate = new Date(model.pullRequest.updated_at);


        const authorLogin = model.pullRequest.user.login;
        this.author = await UserFactory.fetch({
            teamName: model.teamName,
            login: authorLogin,
            isBotUser: model.botUserNames.includes(authorLogin),
            isFormerEmployee: model.formerEmployeeNames.includes(authorLogin)
        });

        return this;
    }

    private initializeDates(model: ImportParams) {
        this.sharedForReviewDate = calculatePrSharedForReviewDate(model);
        this.mergedDate = new Date(model.pullRequest.merged_at);

        const commitTimestamps = model.pullRequestActivities.filter(ActivityTraits.isCommitedEvent).map((c) => new Date(c.author.date).getTime());
        if (commitTimestamps.length > 0) {
            this.initialCommitDate = new Date(Math.min(...commitTimestamps));
            this.lastCommitDate = new Date(Math.max(...commitTimestamps));
        }
        return this;
    }

    private calculateCommitStats(model: ImportParams) {
        this.totalCommentsCount = getCommentsTimestamps(model.pullRequestActivities, model.botUserNames).length;
        this.diffSize = model.files.reduce((acc, file) => acc + file.changes, 0);
        this.testsWereTouched = model.files.some(file => file.filename.toLowerCase().includes("test"));
        return this;
    }

    private async initializeParticipants(model: ImportParams) {
        const allParticipants = new Set<string | null>([
            ...model.pullRequest.requested_reviewers.map(r => r.login),
            ...model.pullRequest.assignees.map(p => p.login),
            ...model.pullRequestActivities.filter(ActivityTraits.isMergedEvent).map(c => c.actor.login),
            ...model.pullRequestActivities.filter(ActivityTraits.isCommentedEvent).map(c => c.actor.login),
            ...model.pullRequestActivities.filter(ActivityTraits.isLineCommentedEvent).flatMap(c => c.comments).map(c => c.user?.login),
            ...model.pullRequestActivities.filter(ActivityTraits.isReviewedEvent).map(c => c.user?.login || null)
        ]);

        this.participants = await Promise.all(
            Array.from(allParticipants)
                .filter((participantLogin): participantLogin is string => !!participantLogin)
                .filter(participantLogin => participantLogin !== model.pullRequest.user.login)
                .map(async participantLogin => {

                    const participantUser = await UserFactory.fetch({
                        teamName: model.teamName,
                        login: participantLogin,
                        isBotUser: model.botUserNames.includes(participantLogin),
                        isFormerEmployee: model.formerEmployeeNames.includes(participantLogin)
                    });

                    return new GitHubPullRequestParticipant(
                        model.teamName,
                        model.pullRequest,
                        getActivitiesOf(model.pullRequestActivities, participantLogin),
                        participantUser
                    );
                }));
        return this;
    }
}
