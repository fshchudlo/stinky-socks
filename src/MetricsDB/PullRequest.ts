import { Column, Entity, JoinColumn, ManyToOne, OneToMany, PrimaryColumn } from "typeorm";
import { PullRequestParticipant } from "./PullRequestParticipant";
import { Contributor } from "./Contributor";

@Entity()
export abstract class PullRequest {
    @Column()
    protected teamName: string;

    @PrimaryColumn()
    protected projectName: string;

    @PrimaryColumn()
    protected repositoryName: string;

    @PrimaryColumn()
    protected pullRequestNumber: number;

    @Column()
    protected viewURL: string;

    @Column()
    protected targetBranch: string;

    @Column()
    protected createdDate: Date;

    @Column()
    protected updatedDate: Date;

    @Column()
    protected sharedForReviewDate: Date;

    @Column({ nullable: true })
    protected initialCommitDate?: Date;

    @Column()
    protected lastCommitDate: Date;

    @Column()
    protected mergedDate: Date;

    @Column()
    protected reviewersCount: number;

    @Column()
    protected commentsCount: number;

    @Column({ type: "numeric" })
    protected diffSize: number;

    @Column()
    protected testsWereTouched: boolean;

    @Column({ type: "varchar" })
    protected authorRole: "COLLABORATOR" | "CONTRIBUTOR" | "FIRST_TIMER" | "FIRST_TIME_CONTRIBUTOR" | "MEMBER" | "OWNER" | "MANNEQUIN" | "NONE";

    @OneToMany(() => PullRequestParticipant, (participant) => participant.pullRequest, {
        cascade: true
    })
    participants: PullRequestParticipant[];

    @ManyToOne(() => Contributor, (author) => author.pullRequests, { onDelete: "CASCADE" })
    @JoinColumn()
    author: Contributor;
}
