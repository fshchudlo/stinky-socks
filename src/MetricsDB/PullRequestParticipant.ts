import { Entity, Column, ManyToOne, JoinColumn, PrimaryColumn } from "typeorm";
import { PullRequest } from "./PullRequest";
import { Contributor } from "./Contributor";

@Entity()
export abstract class PullRequestParticipant {
    @PrimaryColumn()
    protected teamName: string;

    @PrimaryColumn()
    protected projectName: string;

    @PrimaryColumn()
    protected repositoryName: string;

    @PrimaryColumn()
    protected pullRequestNumber: number;

    @PrimaryColumn()
    //Without this typeorm makes really strange things when doing upsert logic
    protected participantIdForPrimaryKeyHack: number;

    @Column({ nullable: true })
    protected firstCommentDate?: Date;

    @Column({ nullable: true })
    protected firstApprovalDate?: Date;

    @Column({ nullable: true })
    protected lastCommentDate?: Date;

    @Column({ nullable: true })
    protected lastApprovalDate?: Date;

    @Column()
    protected commentsCount: number;

    @ManyToOne(() => PullRequest, (pr) => pr.participants, { onDelete: "CASCADE" })
    // ⚠️ remove snake naming after this bugfix merge: https://github.com/typeorm/typeorm/pull/11062
    @JoinColumn([
        { name: "project_name", referencedColumnName: "projectName" },
        { name: "repository_name", referencedColumnName: "repositoryName" },
        { name: "pull_request_number", referencedColumnName: "pullRequestNumber" }
    ])
    pullRequest: PullRequest;

    @ManyToOne(() => Contributor, (participant) => participant.participations, { onDelete: "CASCADE" })
    @JoinColumn()
    participant: Contributor;
}
  