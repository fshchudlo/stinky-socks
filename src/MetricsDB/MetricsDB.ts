import { DataSource } from "typeorm";
import { AppConfig } from "../app.config";
import { PullRequest } from "./entities/PullRequest";
import { PullRequestParticipant } from "./entities/PullRequestParticipant";
import { SnakeNamingStrategy } from "typeorm-naming-strategies";
import { Actor } from "./entities/Actor";

class MetricsDataSource extends DataSource {
    constructor() {
        super({
            type: "postgres",
            host: AppConfig.MetricsDB.DB_HOST,
            port: AppConfig.MetricsDB.DB_PORT,
            username: AppConfig.MetricsDB.DB_USERNAME,
            password: AppConfig.MetricsDB.DB_PASSWORD,
            database: AppConfig.MetricsDB.DB_NAME,
            entities: [PullRequest, PullRequestParticipant, Actor],
            namingStrategy: new SnakeNamingStrategy(),
            synchronize: false,
            migrations: ["src/MetricsDB/migrations/*.ts"],
            migrationsRun: true,
            logging: false
        });
    }

    async getPRsMaxDate(dateFieldName: "mergedDate" | "updatedDate", teamName: string, projectName: string, repositoryName: string) {
        return (await this.getRepository(PullRequest)
            .createQueryBuilder("pr")
            .select(`MAX(pr.${dateFieldName})`, "maxDate")
            .where("pr.teamName = :teamName", { teamName })
            .andWhere("pr.projectName = :projectName", { projectName })
            .andWhere("pr.repositoryName = :repositoryName", { repositoryName })
            .getRawOne())?.maxDate || null;
    }

    async getPRsCount(teamName: string, projectName: string, repositoryName: string): Promise<number> {
        return (await this.getRepository(PullRequest)
            .createQueryBuilder("pr")
            .select(`COUNT(*)`, "prsCount")
            .where("pr.teamName = :teamName", { teamName })
            .andWhere("pr.projectName = :projectName", { projectName })
            .andWhere("pr.repositoryName = :repositoryName", { repositoryName })
            .getRawOne())?.prsCount;
    }
}

export const MetricsDB = new MetricsDataSource();