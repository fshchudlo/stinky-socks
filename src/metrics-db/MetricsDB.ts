import { DataSource } from "typeorm";
import { AppConfig } from "../app.config";
import { PullRequest } from "./PullRequest";
import { PullRequestParticipant } from "./PullRequestParticipant";
import { SnakeNamingStrategy } from "typeorm-naming-strategies";

export const MetricsDB = new DataSource({
    type: "postgres",
    host: AppConfig.MetricsDB.DB_HOST,
    port: AppConfig.MetricsDB.DB_PORT,
    username: AppConfig.MetricsDB.DB_USERNAME,
    password: AppConfig.MetricsDB.DB_PASSWORD,
    database: AppConfig.MetricsDB.DB_NAME,
    entities: [PullRequest, PullRequestParticipant],
    namingStrategy: new SnakeNamingStrategy(),
    synchronize: false,
    migrations: ["src/metrics-db/migrations/*.ts"],
    migrationsRun: true,
    logging: false
});