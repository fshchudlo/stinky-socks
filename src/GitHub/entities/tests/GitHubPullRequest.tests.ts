import { GitHubPullRequest } from "../GitHubPullRequest";
import { ActorFactory } from "../../../MetricsDB/ActorFactory";
import { TestPRsBuilder } from "./TestPRsBuilder";

describe("GitHubPullRequest", () => {
    it("Should map basic fields", async () => {
        jest.spyOn(ActorFactory, "fetch").mockImplementation(async ({ teamName, login, isBotUser }: {
            teamName: string,
            login: string,
            isBotUser: boolean
        }) => {
            return Promise.resolve({
                id: 1,
                login: login,
                teamName: teamName,
                nickname: `${login} nickname`,
                isBotUser: isBotUser,
                isFormerParticipant: false,
                duplicates: [],
                participations: [],
                pullRequests:[],
                teamRole: null
            });
        });
        const model = new TestPRsBuilder().pullRequest();

        const pullRequestEntity = await new GitHubPullRequest().init(model);
        expect(pullRequestEntity).toMatchSnapshot();
    });
});

