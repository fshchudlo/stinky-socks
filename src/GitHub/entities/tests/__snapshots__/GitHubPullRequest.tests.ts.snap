// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubPullRequest Should map basic fields 1`] = `
GitHubPullRequest {
  "author": {
    "duplicates": [],
    "id": 1,
    "isBotUser": false,
    "isFormerParticipant": false,
    "login": "TestUser",
    "nickname": "TestUser nickname",
    "participations": [],
    "pullRequests": [],
    "teamName": "Test team",
    "teamRole": null,
  },
  "authorRole": "MEMBER",
  "createdDate": 2021-01-01T00:00:00.000Z,
  "diffRowsAdded": 0,
  "diffRowsDeleted": 0,
  "firstReactionDate": null,
  "initialCommitDate": undefined,
  "integrityErrors": undefined,
  "lastApprovalDate": null,
  "lastCommitDate": undefined,
  "mergedDate": 2021-01-01T00:03:00.000Z,
  "participants": [],
  "projectName": "TestOwner",
  "pullRequestNumber": 1,
  "repositoryName": "TestRepo",
  "requestedReviewersCount": 0,
  "reworkCompletedDate": null,
  "sharedForReviewDate": 2021-01-01T00:00:00.000Z,
  "targetBranch": "main",
  "teamName": "Test team",
  "testsWereTouched": false,
  "totalCommentsCount": 0,
  "updatedDate": 2021-01-01T00:01:00.000Z,
  "viewURL": "https://github.com/TestOwner/TestRepo/pull/1",
}
`;
